### PR TITLE
Add mask to match template demo

### PR DESCRIPTION
**Pull request by acinader**
Accept and use optional mask argument for MatchTemplate_Demo.

Useful for visualizing effect of a mask on template matching.

this replaces #6758  which was giving me some problems with github.... you can close this if you want to re-open #6758 instead?

### CODE REVIEW

**Comment by acinader** [comment]
I can rebase if you prefer, but i think it makes sense to have two commits

File: doc/tutorials/imgproc/histograms/template_matching/template_matching.markdown [hunk 1]
```diff
@@ -19,6 +19,10 @@ Theory
 Template matching is a technique for finding areas of an image that match (are similar) to a
 template image (patch).
 
+While the patch must be a rectangle it may be that not all of the
+rectangle is relevent.  In such a case, a mask can be used to isolate the portion of the patch
```
Comment (apavlenko) [inline doc/tutorials/imgproc/histograms/template_matching/template_matching.markdown]: relevAnt

File: samples/cpp/tutorial_code/Histograms_Matching/MatchTemplate_Demo.cpp [hunk 2]
```diff
@@ -31,15 +32,20 @@ int main( int argc, char** argv )
...
   }
 
   /// Load image and template
   img = imread( argv[1], IMREAD_COLOR );
   templ = imread( argv[2], IMREAD_COLOR );
 
-  if(img.empty() || templ.empty())
+  if(argc > 3) {
+    use_mask = true;
+    mask = imread(argv[3], IMREAD_COLOR);
```
Comment (apavlenko) [inline samples/cpp/tutorial_code/Histograms_Matching/MatchTemplate_Demo.cpp]: it's stated that "The mask should be a greyscale image", but here it's forced to 3 channels

File: doc/tutorials/imgproc/histograms/template_matching/template_matching.markdown [hunk 3]
```diff
@@ -150,10 +177,14 @@ Explanation
     @endcode
 -#  Perform the template matching operation:
     @code{.cpp}
-    matchTemplate( img, templ, result, match_method );
+    bool method_accepts_mask = CV_TM_SQDIFF == match_method || match_method == CV_TM_CCORR_NORMED;
```
Comment (apavlenko) [inline doc/tutorials/imgproc/histograms/template_matching/template_matching.markdown]: I suggest similar order in both comparisons:
```cpp
bool method_accepts_mask = (CV_TM_SQDIFF == match_method || CV_TM_CCORR_NORMED == match_method);
```

File: doc/tutorials/imgproc/histograms/template_matching/template_matching.markdown [hunk 4]
```diff
@@ -115,8 +142,8 @@ Explanation
     @endcode
 -#  Load the source image and template:
     @code{.cpp}
-    img = imread( argv[1], 1 );
-    templ = imread( argv[2], 1 );
+    img = imread( argv[1], IMREAD_COLOR );
+    templ = imread( argv[2], IMREAD_COLOR );
     @endcode
```
Comment (apavlenko) [inline doc/tutorials/imgproc/histograms/template_matching/template_matching.markdown]: shouldn't mask load be mentioned here?

File: doc/tutorials/imgproc/histograms/template_matching/template_matching.markdown (hunk 1 above)
Comment (acinader) [inline doc/tutorials/imgproc/histograms/template_matching/template_matching.markdown]: ugh.  installing spell checker.  thanks.

**Comment by acinader** [comment]
@apacha thanks for the feedback.  I think i've covered all your issues in b53b8d9

**Comment by snosov1** [comment]
Although, it's not necessarily the issue of this PR, but a wording in the documentation made me wonder about the implementation. The statement that took my attention was:

> A value between black and white will include some of the match in proportion to how dark the pixel is.

1. According to the implementation this statement is _incorrect_ for masks of `CV_8U` depth, see https://github.com/opencv/opencv/blob/master/modules/imgproc/src/templmatch.cpp#L860 . The `compare` call makes all non-zero values equal to 255, the consequent conversion makes them equal to 1.
2. Then, for `CV_32F` masks this statement seems to hold (which is already strange, since the 2 options don't work the same). But looking at the implementation, I don't think it's implemented correctly (or I just don't get it):
    1. For `CV_TM_CCORR_NORMED` metric, I would expect that the implementation will simply multiply the template by mask. But it does something more involved.
    2. For `CV_TM_SQDIFF` metric, I would expect it to multiply the squared bracket by mask. It looks like the implementation tries to replace the square of sum into separate units and multiply them by mask individually. But to me, the evaluation doesn't look correct.

So, @acinader are you familiar with the implementation and can you confirm whether it works correctly?

Also, it looks like @wangyan42164 is the author of the implementation. Is that so? If yes, can you, please, comment as well?

**Comment by acinader** [comment]
@snosov1 I am not familiar with the implementation.  I am coming at it as a user and trying to figure out how templates and masks work by trial and error (which is why I am adding the mask to the sample so I can play and see the results.)

FWIW, it would be helpful to have some docs and unit tests with some sample images that could be a guide for how to prepare a template and mask.  Unfortunately, I don't have enough understanding of what is going on to drive that process, but I'd be glad to help.

**Comment by snosov1** [comment]
Ok, then. I think we can merge this particular PR after a couple of small changes in wording. However, we'll probably need to submit a separate ticket about the issues I highlighted (different behavior for `CV_8U` and `CV_32F`, correctness of implementation).

So, I suggest that you replace the following paragraph with the proposed text, and if you agree with it - please, update the PR and we'll merge it.

> The mask should be a grayscale image where each pixel contains some value from black to white.
> Pixels that are white are fully included in calculating the best match.  Pixels that are black
> are excluded from the match.  A value between black and white will include some of
> the match in proportion to how dark the pixel is.  Although the image should be a grayscale whose
> output from the file command should look something like: "PNG image data, 128 x 128, 8-bit gray
> +alpha, non-interlaced", opencv will read the image into an rgb matrix that will be applied
> during the image match.

The mask should have a `CV_8U` or `CV_32F` depth and the same number of channels
as the template image. In `CV_8U` case, the mask values are treated as binary,
i.e. zero and non-zero. In `CV_32F` case, the values should fall into `[0..1]`
range and the template pixels will be multiplied by the corresponding mask pixel
values. Since the input images in the sample have the `CV_8UC3` type, the mask
is also read as color image.

**Comment by acinader** [comment]
@snosov1 updated to reflect your suggestion.  FYI, your explanation is very helpful to me, so a good change!

thanks.

**Comment by snosov1** [comment]
@apavlenko, I'm ok with this PR.

I'll create a separate issue with my concerns about the implementation.

**Comment by apavlenko** [comment]
:+1:

**Comment by Cartucho** [comment]
@acinader could you send me the images you used for the mask part? 
I'm trying to do a Python and Java version on this

The images I need are these 3:
screenshot from 2017-02-19 18-11-03

**Comment by sturkmen72** [comment]
@Cartucho you can find them in samples/data folder

https://github.com/opencv/opencv/blob/master/samples/data/lena_tmpl.jpg
https://github.com/opencv/opencv/blob/master/samples/data/tmpl.png
https://github.com/opencv/opencv/blob/master/samples/data/mask.png

**Comment by Cartucho** [comment]
Thanks! @sturkmen72

**Comment by ai-snow** [comment]
@Cartucho Did you get this to work?  When I run these images through the matchTemplate function I get incorrect results:

image

**Comment by Cartucho** [comment]
@ai-snow Did you have a look at this [tutorial](http://docs.opencv.org/trunk/de/da9/tutorial_template_matching.html)?

**Comment by ai-snow** [comment]
@cartucho

Thanks for referring to that document. However, this does not change my opinion that the implementation in OpenCV for match templates with a mask is incorrect. (I'm not asking a question about my implementation in my comment; that was there for completeness)

**Comment by ai-snow** [comment]
@cartucho
Did you get it to work in python? I used that tutorial as a starting point. Seemed pretty straightforward in terms of seeing it up.

**Comment by Cartucho** [comment]
@ai-snow Shouldn't you use  `res = cv2.matchTemplate(im_main, im_temp, cv2.TM_SQDIFF, None, im_mask)
` instead of  `res = cv2.matchTemplate(im_main, im_temp, cv2.TM_SQDIFF, im_mask)`

**Comment by ai-snow** [comment]
@cartucho

Wow, where did you find that method signature? That seems to work, but I'm afraid with all the searching I've done, that hasn't popped up anywhere. Thanks for pointing this out!

**Comment by Cartucho** [comment]
@ai-snow I did this PR #9260  that will be adding the Python signatures to the functions docs! I still need to fix some things, but in a couple of weeks it should be merged!

**Comment by alalek** [comment]
@ai-snow Try this `python2 -c 'import cv2; help(cv2.matchTemplate)'` to get signature

**Pull request merged**